event/system.go

```diff
@@ -0,0 +1,170 @@
+package event
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Subscriber is a limited interface of System that only allows subscribing and
+// unsubscribing.
+type Subscriber interface {
+	Subscribe(events ...Type) (subID uint64, eventsCh <-chan *Event)
+	Unsubscribe(subID uint64)
+}
+
+// System keeps track of subscribers, and allows subscribing to and emitting
+// events.
+type System struct {
+	subMx       sync.RWMutex
+	subIDCount  uint64
+	subscribers map[Type]map[uint64]chan *Event
+	eventBuffer int
+	logger      logrus.FieldLogger
+}
+
+// NewEventSystem returns a new System.
+// eventBuffer determines the size of the Event channel buffer. Events might be
+// dropped if this buffer is full and there are no event listeners, or if events
+// are emitted very quickly and the event handler goroutine is busy. It is
+// recommended to handle events in a separate goroutine to not block the
+// listener goroutine.
+func NewEventSystem(eventBuffer int, logger logrus.FieldLogger) *System {
+	return &System{
+		subscribers: make(map[Type]map[uint64]chan *Event),
+		eventBuffer: eventBuffer,
+		logger:      logger,
+	}
+}
+
+// Subscribe to one or more events. It returns a subscriber ID that can be
+// used to unsubscribe, and an Event channel to receive events.
+// It panics if events is empty.
+func (s *System) Subscribe(events ...Type) (subID uint64, eventsCh <-chan *Event) {
+	if len(events) == 0 {
+		panic("must subscribe to at least 1 event type")
```

Comment:
Because this is a programming error. Calling `Subscribe()` with 0 events is invalid. I'm only making it fail explicitly to avoid unnecessary locking, and everything that happens below.

Reply:
That's correct, but I still missing why this is a panic and not an error, which also could prevent everything that happens bellow.

TL;DR, I'm asking why not:
```diff
- func (s *System) Subscribe(events ...Type) (subID uint64, eventsCh <-chan *Event) {
- 	if len(events) == 0 {
-		panic("must subscribe to at least 1 event type")
+ func (s *System) Subscribe(events ...Type) (subID uint64, eventsCh <-chan *Event) error {
+	if len(events) == 0 {
+		return errors.New("must subscribe to at least 1 event type")
```

Reply:
It might be similar to [this](https://cs.opensource.google/go/go/+/refs/tags/go1.20.5:src/strings/replace.go;l=26-37). To keep the API more straightforward to use.

I've been just following this PR from time to time. You can ignore me :)

Reply:
Exactly, thank you Inanc :bow:

This is a programming, and not an application or user error. So returning `error` instead would just be annoying to use.

If Go allowed a build-time check for this case, I would prefer that instead, but alas. :disappointed:

Reply:
I'm still of the opinion that there could be cases where a user of this API is deciding if they should or not should process after catching of error from there. And catching the panic isn't the best pattern vs. just an error. But keeping in mind the risk of happening such cases prefer just to resolve the tread :sweat: